network/connPerfMon.go

```diff
@@ -384,3 +385,127 @@ func (pm *connectionPerformanceMonitor) accumulateMessage(msg *IncomingMessage,
 		delete(msgBucket.messages, msgDigest)
 	}
 }
+
+type networkAdvanceMonitor struct {
+	// lastNetworkAdvance contains the last timestamp where the agreement protocol was able to make a notable progress.
+	// it used as a watchdog to help us detect connectivity issues ( such as cliques )
+	lastNetworkAdvance time.Time
+
+	mu deadlock.Mutex
+}
+
+func makeNetworkAdvanceMonitor() *networkAdvanceMonitor {
+	return &networkAdvanceMonitor{
+		lastNetworkAdvance: time.Now().UTC(),
+	}
+}
+
+func (m *networkAdvanceMonitor) lastAdvancedWithin(interval time.Duration) bool {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	// now < last + interval <=> now - last < interval
+	return time.Now().UTC().Before(m.lastNetworkAdvance.Add(interval))
+}
+
+func (m *networkAdvanceMonitor) updateLastAdvance() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.lastNetworkAdvance = time.Now().UTC()
+}
+
+type outgoingConnsCloser struct {
+	log                   logging.Logger
+	net                   outgoingDisconnectable
+	cliqueResolveInterval time.Duration
+	connPerfMonitor       *connectionPerformanceMonitor
+	netAdvMonitor         *networkAdvanceMonitor
+}
+
+type outgoingDisconnectable interface {
+	outgoingPeers() (peers []Peer)
+	numOutgoingPending() int
+	disconnect(badnode Peer, reason disconnectReason)
+	OnNetworkAdvance()
+}
+
+func makeOutgoingConnsCloser(log logging.Logger, net outgoingDisconnectable, connPerfMonitor *connectionPerformanceMonitor, cliqueResolveInterval time.Duration) *outgoingConnsCloser {
+	return &outgoingConnsCloser{
+		log:                   log,
+		net:                   net,
+		cliqueResolveInterval: cliqueResolveInterval,
+		connPerfMonitor:       connPerfMonitor,
+		netAdvMonitor:         makeNetworkAdvanceMonitor(),
+	}
+}
+
+// checkExistingConnectionsNeedDisconnecting check to see if existing connection need to be dropped due to
+// performance issues and/or network being stalled.
+func (cc *outgoingConnsCloser) checkExistingConnectionsNeedDisconnecting(targetConnCount int) bool {
+	// we already connected ( or connecting.. ) to  GossipFanout peers.
+	// get the actual peers.
+	outgoingPeers := cc.net.outgoingPeers()
+	if len(outgoingPeers) < targetConnCount {
+		// reset the performance monitor.
+		cc.connPerfMonitor.Reset([]Peer{})
+		return cc.checkNetworkAdvanceDisconnect()
+	}
+
+	if !cc.connPerfMonitor.ComparePeers(outgoingPeers) {
+		// different set of peers. restart monitoring.
+		cc.connPerfMonitor.Reset(outgoingPeers)
```

Comment:
@cce `outgoingPeers` is a copy of `wsPeers` i.e. only peers maintained by p2p net, not any extra peers from pubsub (at least for hybrid where we check connection source). For non-hybrid p2p it will also have pubsub discovered peers but it does not matter for this PR as soon as we aim to limit conns for hybrid mode